ドキュメント詳細ページを作成

kazuma-naka opened the pull request:
## Issue
#18 

## 概要
- `DocList.tsx` でタイトルをクリックすると `docs/[id]` に画面遷移する

~~- `src/app/docs/[id]/page.tsx` に ` "title, body, user_id, last_updated_user_id, created_at, updated_at"` を取得して表示する~~
~~-  `src/app/docs/[id]/page.tsx` に ` "body, user_name, created_at, updated_at"` を取得して表示する
     - user_id を使ってユーザー情報を取得しようとしましたが `getUserById()` を使うためには `admin` 権限が必要であったためユーザーの名前をドキュメント作成時に `string` 型で登録する設計にしました。~~

- /docs/[id] の parameter から docs の id を取得し `"title, body, user_id, created_at, updated_at"` を取得して表示する。
- user_id から last_name を取得して表示
- /docs/[id]/page.tsx はサーバーコンポネントなので `user.role` を取得できない。 `supabase/ssr` の `createServerClient()` を使って `public.users` の `auth.role() = 'authenticated'` の RLS ポリシーに違反しないようにする。

-  E2E テストを作成

<details>
<summary> １回目のスクショ </summary>

![doc-details](https://github.com/user-attachments/assets/d6421f6b-ecf6-49b8-bf7d-f68472d30a1b)
</details>

<details>
![doc_detail_photo](https://github.com/user-attachments/assets/58422217-3bba-416c-98a3-44a9f5c1a7d6)
</details>

### ログアウト時：ユーザー名は表示されない
![details_2](https://github.com/user-attachments/assets/a3da2985-dd62-4700-ac2e-fff158bb47f7)

### ログイン時：ユーザー名が表示される
![details_login](https://github.com/user-attachments/assets/76b8cad5-2845-422b-acbf-2c491d14c422)

## signup.test.ts の修正

- `waitForEmail()` でEmail アドレスからメールの一覧を取得する。

- `getEmailDetails()` で id から本文を取得してURLリンクを取得する

kazuma-naka committed:
DocList.tsx のコンポネントテストを作成

モックを作成する

kazuma-naka committed:
Doc の詳細ページを作成。 id を表示

kazuma-naka committed:
Dynamic に作成した詳細 Page に渡された id で title, body, user_id. last_updated_user_id, created_at, updated_at を取得する

kazuma-naka committed:
docs/[id] となるように修正

kazuma-naka committed:
ドキュメント詳細のE2Eテストを作成する

kazuma-naka committed:
ドキュメント詳細のE2Eテストですべての値が表示されるか修正する

kazuma-naka committed:
ドキュメント詳細のE2Eテストですべての値が表示されるか修正する

kazuma-naka committed:
ドキュメント削除ボタンをドキュメント詳細画面に追加

kazuma-naka committed:
E2E テストを追加する

m0nch1 commented on playwright.config.ts:
```diff
@@ -3,44 +3,34 @@ import { defineConfig, devices } from "@playwright/test";
-    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
     trace: "on-first-retry",
     headless: true,
+    timezoneId: "UTC",
```
利用者が日本向けだと思うので JST で良いかなぁと思ったんですがどうでしょう！

kazuma-naka replied:
JST に変更しました！

kazuma-naka replied:
JST に変更しました！

m0nch1 commented on src/app/docs/[id]/page.tsx:
```diff
@@ -0,0 +1,37 @@
+      <h1>{doc.title}</h1>
+      <p>{doc.body}</p>
+      <p>
+        <strong>User ID:</strong> {doc.user_id}
```
もし、太字にしたいだけならば strong タグは適切でないと思います！ css を当てましょう！

kazuma-naka replied:
css を使用しました

m0nch1 commented on e2e/doclist.test.ts:
```diff
@@ -4,21 +4,49 @@ import { withSupawright } from "supawright";
 import type { Database } from "@/lib/database.types";
 
 const test = withSupawright<Database, "public">(["public"]);
+const currentDate = "2025-02-08T23:00:38.455Z";
```
https://playwright.dev/docs/clock API を使うと良さそうかなと思ったんですがどうでしょう！

kazuma-naka replied:
toBeVisible を使用して細かく時間を指定しないように修正しました

m0nch1 commented on e2e/doclist.test.ts:
[q]

e2e をどの単位でテストしていくのかちょっと気になりポイントです！
page 単位で作っていくのがいいんじゃないかな〜と思うんですがどうでしょう！ 

ref: https://kentcdodds.com/blog/the-testing-trophy-and-testing-classifications

kazuma-naka replied:
ドキュメント一覧はコンポネントテストがあり十分だと考えたのでドキュメント詳細ページのE2Eテストのみ残しました。

m0nch1 replied:
ファイル名なんですが、 doclist だとコンポーネントのテストかなぁって感じがしますね！
今回の場合、 e2e/docs/[id]/page.spec.ts みたいなものをイメージしてました！

---
というか、もしかしたら https://nextjs.org/docs/app/getting-started/project-structure#private-folders のように、app/docs/[id]/_test/page.e2e.test.ts みたいにするといいのかもとも思ってきました。

これはこの PR では対応しなくて良いと思ってるんですがどこかでフォルダ構成見直せると良いかなと思いました！

m0nch1 reviewed (commented):
@kazuma-naka 

issue では

```
本文
作成者
createdAt
updatedAt
```

とあり、この PR では `title, body, user_id, last_updated_user_id, created_at, updated_at` を表示しているようですが、last_updated_user_id と title は必要ですか？

また、作成者は id を表示するだけで良さそうでしょうか！

kazuma-naka committed:
JST に変更

kazuma-naka committed:
strong タグではなく css を使う

kazuma-naka committed:
E2Eテストを必要なもののみにする

ドキュメント詳細ページに遷移して必要なデータが表示されているかテストする

kazuma-naka committed:
createdAt と updatedAt の static な時間を削除

kazuma-naka committed:
last_updated_user_id と title を表示しない

kazuma-naka committed:
doc の user_id を UUID にする。users テーブルから名前を取得する

kazuma-naka committed:
doc の user_id を user_name に変更する

kazuma-naka committed:
E2E テストを修正

User Id: から User: に変更

kazuma-naka commented:
> @kazuma-naka
> 
> issue では
> 
> ```
> 本文
> 作成者
> createdAt
> updatedAt
> ```
> 
> とあり、この PR では `title, body, user_id, last_updated_user_id, created_at, updated_at` を表示しているようですが、last_updated_user_id と title は必要ですか？
> 
> また、作成者は id を表示するだけで良さそうでしょうか！

`body, user_name, created_at, updated_at` のみを表示するように修正しました。
作成者の `id` からユーザー情報を取得できるかと思っていたのですがユーザーを id から取得する `getUserById()` を使うには admin 権限が必要で `service role key` が必要でした。
ですのでドキュメント作成時にユーザーの名前を `string` で登録する方向性で修正しました🖊

kazuma-naka committed:
updated_at を更新する

m0nch1 commented:
@kazuma-naka 
https://github.com/fjordllc/agent/pull/54#discussion_r1949198807 だけ確認してもらえれば、あとは良さそうに思いました！

kazuma-naka committed:
pagination を Chakra UI を使って作成する

kazuma-naka committed:
pagination を chakra UI を使って実装

kazuma-naka committed:
ディレクトリやnits なものの修正

kazuma-naka committed:
docs の引数を user_id, last_updated_user_id に変更

kazuma-naka committed:
DocList から DocDetails に遷移できるようにする

details ページの params を Promise に変更する

kazuma-naka committed:
DB から docs の body, user_id, created_at, updated_at のみを取得する

kazuma-naka committed:
DocDetails に shadcn の Card を使用する

kazuma-naka committed:
コンポネントテストのモックを next/navigation を追加する

useRoute で詳細ページに遷移するため

kazuma-naka committed:
DocList の詳細ページに遷移する E2E テストのディレクトリを修正

kazuma-naka committed:
public.users から user_id を使って last_name, first_name を取得しようとしたが RLS を Diabled にしないといけない

kazuma-naka committed:
public.users から user_id を使って last_name, first_name を取得しようとしたが RLS を Diabled にしないといけない

kazuma-naka committed:
DocDetails はサーバーコンポネントなので auth.getuser が supabase-js の createClient() では取得できない。なので supabaseServerClient を作成する

kazuma-naka committed:
e2e テストを移動する

kazuma-naka commented:
@m0nch1 さん、

こちらの PR も修正しましたので確認よろしくお願いします。
`user_id` から `public.users` の `last_name` を取得してドキュメント詳細ページに表示する設計にしました。
`docs/[id]/page.tsx` はサーバーコンポネントなため、 supabase/ssr を使い auth.getUser() で role を取得できるようにしました。
public.users は RLS により auth.role() = 'authenticated' でないと Select できないからです。
そのため ./src/lib/supabaseServer.ts を新たに作成しました。

kazuma-naka committed:
supabaseServer.ts を公式ドキュメントと同じにする

kazuma-naka committed:
DeleteButton を BootCamp の削除ボタンと似るように修正

kazuma-naka committed:
e2e テストを serial で run する

kazuma-naka committed:
内容修正ボタンの修正

kazuma-naka committed:
doc details を shadcn を使って修正する

kazuma-naka committed:
不必要なフックを削除

kazuma-naka committed:
docs/[id]/edit/page をサーバーコンポネントにする

kazuma-naka committed:
e2e/docs/[id] の配下に E2E テストファイルを配置

kazuma-naka committed:
DocDeleteButton を docs/_compoents に移動。router.replace(/docs) に変更してもドルをクリックしたときに削除されたドキュメントに戻らないように修正

kazuma-naka committed:
Link を replace に変えて戻るボタンが押されたときにedit 画面に戻らないようにする

kazuma-naka committed:
EditDoc の params を Promise に変更する

Next.js v15 のため

kazuma-naka committed:
Doc作成ボタンを作成

kazuma-naka committed:
doc/new を新たに作成

kazuma-naka committed:
ドキュメント新規作成のE2E テストを作成

kazuma-naka committed:
cast を使わずエラーハンドリングする

kazuma-naka committed:
doc/edit/page に変更。searchParams を使用して docs.id を取得する

kazuma-naka committed:
テストの名前を修正

kazuma-naka committed:
cast を無くしてエラーハンドリングに修正

kazuma-naka committed:
docs/new/page.tsx の form の部分を切り出して client component として作成する

kazuma-naka committed:
テストドキュメントを削除後のテストを追加

kazuma-naka committed:
react-markdown v10.0.0 と remark-gfm v4.0.1 をインストール

kazuma-naka committed:
titleRaw から title に変更する

kazuma-naka committed:
@tailwindcss/typography をインストールして <h> タグをマークダウンで表示

kazuma-naka committed:
react-syntax-highlighter をインストールして syntax プラグインを反映する

kazuma-naka committed:
tailwind でドキュメント詳細画面を整える

kazuma-naka committed:
run npm install

kazuma-naka committed:
run npm install

m0nch1 commented on src/lib/supabaseServer.ts:
📝 memo
ref: https://supabase.com/docs/guides/auth/server-side/nextjs

m0nch1 commented on src/lib/supabaseServer.ts:
[nits]

src/lib/client の中にまとめてしまうか、https://supabase.com/docs/guides/auth/server-side/nextjs のように `utils/supabase/server.ts` と `utils/supabase/client.ts` にすると良いかなと思いました！

m0nch1 commented on src/lib/supabase.ts:
```diff
@@ -1,5 +1,5 @@
 import { createClient } from "@supabase/supabase-js";
-import { Database } from "./database.types";
+import { Database } from "../../supabase/database.types";
```
[q] 

src/lib の database.types.ts と、supabase/supabase/database.types.ts って一緒ですかね？？？

なんかまとめられるのであればまとめた方が良さそう！

m0nch1 reviewed (commented):
@kazuma-naka 
ちょっと q コメントあります！！ 🙏

kazuma-naka committed:
Error Toast をクライアントサイドで表示する ClientErrorToaster を作成。ドキュメント詳細ページでドキュメントを作成したユーザーの取得に失敗、ドキュメント取得に失敗した場合、Toast を表示する

kazuma-naka committed:
utils/supabase/server が存在するので lib/supabaseServer を削除する

kazuma-naka committed:
./supabase/database.type.ts を ./src/lib/ 配下に移動して一つに統一する

kazuma-naka committed:
@/utils/supabase/server に path を修正する

kazuma-naka committed:
@/utils/supabase/server に path を変更

kazuma-naka commented on src/app/docs/[id]/page.tsx:
```diff
@@ -0,0 +1,71 @@
+    .maybeSingle();
+
+  if (docError) {
+    console.error(
+      `ドキュメントの取得に失敗しました。\n${docError.code} ${docError.message}`,
```
ClientErrorToaster を作成しエラー Toast を表示するように修正しました。

kazuma-naka committed:
app/docs/page.tsx のコンフリクトを解消

kazuma-naka committed:
docs/[id]/edit に変更する

user_id を使って public.users から last_name を取得する仕様に変更したので e2e テストをテストユーザーでログインするもののみにする

kazuma-naka committed:
正規表現が何をするのかコメントを追加する

kazuma-naka committed:
e2e/docs/new/page.e2e.test.ts がテストユーザーでログイン後、ドキュメントを追加、詳細ページに遷移、ドキュメントの削除を行うものなので重複したE2Eテストを削除する

m0nch1 reviewed (approved):
markdown とかのが入っているが一旦 LGTM

komagata commented:
@kazuma-naka conflictの修正をお願いします。

kazuma-naka committed:
コンフリクトの解消

shadcn のバージョンアップ

kazuma-naka committed:
npm install を実行して package-lock.json ファイルを揃える

kazuma-naka committed:
tailwindcss/typography をインストールする

kazuma-naka committed:
Mailpit で email アドレスからメッセージの一覧を取得、メッセージの id から本文を取得する

kazuma-naka committed:
email ID からメッセージ本文を取得して Confirmation の URL を取得するように修正

kazuma-naka committed:
e2e/docs/new/page.e2e.test.ts のタイポと重複する describe を削除

kazuma-naka commented on e2e/docs/new/page.e2e.test.ts:
```diff
@@ -0,0 +1,57 @@
+
+const test = withSupawright<Database, "public">(["public"]);
+
+describe("test to create new document", () => {
```
重複する describe を削除しました。

komagata reviewed (approved):
確認させて頂きました。OKです〜🙆‍♂️
